core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -242,64 +226,114 @@ void forceRefresh() {
+                TimeUnit.SECONDS);
+      }
+
+      return data;
+    } catch (Exception e) {
+      synchronized (instanceDataGuard) {
+        nextInstanceData = executor.submit(() -> performRefresh());
+      }
+      throw e;
+    }
+  }
+
+  @VisibleForTesting
+  public ListenableFuture<InstanceData> getCurrentInstanceData() {
+    synchronized (instanceDataGuard) {
+      return currentInstanceData;
+    }
+  }
+
+  @VisibleForTesting
+  ListenableFuture<InstanceData> getNextInstanceData() {
+    synchronized (instanceDataGuard) {
+      return nextInstanceData;
+    }
+  }
+
+  private ListenableFuture<InstanceData> fetchInstanceData() {
+    final GoogleCredentials credentials;
     if (iamAuthnCredentials.isPresent()) {
-      GoogleCredentials downscopedCredentials = getDownscopedCredentials(iamAuthnCredentials.get());
-      refreshFuture =
-          apiFetcher.getInstanceData(
-              instanceName, downscopedCredentials, AuthType.IAM, executor, keyPair);
+      credentials = getDownscopedCredentials(iamAuthnCredentials.get());
     } else {
-      refreshFuture =
-          apiFetcher.getInstanceData(instanceName, null, AuthType.PASSWORD, executor, keyPair);
+      credentials = null;
     }
 
-    Futures.addCallback(
-        refreshFuture,
-        new FutureCallback<InstanceData>() {
-          @Override
-          public void onSuccess(InstanceData instanceData) {
-            synchronized (instanceDataGuard) {
-              // update currentInstanceData with the most recent results
-              currentInstanceData = refreshFuture;
-              // schedule a replacement before the SSLContext expires;
-              nextInstanceData =
-                  executor.schedule(
-                      () -> performRefresh(),
-                      secondsUntilRefresh(getInstanceData().getExpiration()),
-                      TimeUnit.SECONDS);
-            }
-          }
-
-          @Override
-          public void onFailure(Throwable t) {
-            logger.log(
-                Level.WARNING,
-                "An error occurred while performing refresh. Retrying immediately.",
-                t);
-            synchronized (instanceDataGuard) {
-              InstanceData instanceData = null;
-              try {
-                instanceData = getInstanceData();
-              } catch (Exception e) {
-                // this means the result was invalid
-              }
-              if (instanceData == null
-                  || instanceData.getExpiration().toInstant().isBefore(Instant.now())) {
-                // replace current if it is expired or invalid
-                currentInstanceData = refreshFuture;
+    // Fetch the metadata
+    ListenableFuture<Metadata> metadataFuture =
+        executor.submit(() -> apiFetcher.fetchMetadata(instanceName, authType));
+
+    // Fetch the ephemeral certificates
+    ListenableFuture<Certificate> ephemeralCertificateFuture =
+        Futures.whenAllComplete(keyPair)
```

Comment:
`Futures.whenAllComplete()` will throw an exception if a future fails. Previously this used `Futures.whenAllSucceeded()` which will hang if a future never succeeds.

Reply:
Should this land as a single PR?